Profiler: add Self CPU Time Total, CPU time total and other general improvements

salexspb opened the pull request:
Summary:
Function profile events are typically nested. In this diff I
add parent child relationship to the intervals. This way we can
attribute self time easily. As a result, user printing a table from a
profiler trace gets self cpu time.

This diff doesn't try to address CUDA self time as CUDA kernels are
already getting special care in the profiler.

There are also some other minor improvements. Like reporting total CPU
time spent, reversed sorting, aggregated data after the table,
etc.

There is a new unit test added which tests more functionality than
previous profiler test

```
test/autograd -r profiler_agg
test_profiler_aggregation (test_autograd.TestAutograd) ... -----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
Name                     Self CPU total %   Self CPU total      CPU total %        CPU total     CPU time avg     CUDA total %       CUDA total    CUDA time avg  Number of Calls
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
lstm                               0.44%          2.502ms            7.16%         40.790ms         40.790ms              inf          0.000us          0.000us                1
lstm                               0.35%          1.987ms            5.53%         31.516ms         31.516ms              inf          0.000us          0.000us                1
lstm                               0.35%          1.968ms            5.47%         31.175ms         31.175ms              inf          0.000us          0.000us                1
lstm                               0.33%          1.863ms            5.03%         28.684ms         28.684ms              inf          0.000us          0.000us                1
lstm                               0.32%          1.837ms            5.03%         28.682ms         28.682ms              inf          0.000us          0.000us                1
lstm                               0.31%          1.789ms            4.94%         28.143ms         28.143ms              inf          0.000us          0.000us                1
lstm                               0.31%          1.778ms            4.87%         27.768ms         27.768ms              inf          0.000us          0.000us                1
lstm                               0.31%          1.776ms            4.89%         27.842ms         27.842ms              inf          0.000us          0.000us                1
lstm                               0.31%          1.743ms            4.82%         27.453ms         27.453ms              inf          0.000us          0.000us                1
lstm                               0.31%          1.742ms            4.69%         26.744ms         26.744ms              inf          0.000us          0.000us                1
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
Self CPU time total: 569.819ms
CUDA time total: 0.000us

-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
Name                     Self CPU total %   Self CPU total      CPU total %        CPU total     CPU time avg     CUDA total %       CUDA total    CUDA time avg  Number of Calls
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
sigmoid                           17.29%         98.528ms           17.29%         98.528ms        164.213us              inf          0.000us          0.000us              600
mul                               17.26%         98.378ms           17.26%         98.378ms        163.963us              inf          0.000us          0.000us              600
addmm                             16.46%         93.771ms           16.46%         93.771ms        234.427us              inf          0.000us          0.000us              400
add                               11.80%         67.225ms           11.80%         67.225ms        168.063us              inf          0.000us          0.000us              400
tanh                              10.83%         61.721ms           10.83%         61.721ms        154.303us              inf          0.000us          0.000us              400
split                              6.79%         38.688ms            6.79%         38.688ms        193.441us              inf          0.000us          0.000us              200
lstm                               6.29%         35.857ms          100.00%        569.819ms         28.491ms              inf          0.000us          0.000us               20
unsigned short                     5.41%         30.842ms            5.41%         30.842ms         77.104us              inf          0.000us          0.000us              400
stack                              3.39%         19.322ms            3.39%         19.322ms        241.524us              inf          0.000us          0.000us               80
unbind                             2.34%         13.334ms            2.34%         13.334ms        166.677us              inf          0.000us          0.000us               80
-----------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
Self CPU time total: 569.819ms
CUDA time total: 0.000us

Total time based on python measurements:  589.088ms
CPU time measurement python side overhead: 3.38%
ok
```

Differential Revision: D14988612

soumith commented:
cool diff. what does the profiler output look like now after this diff?

ezyang commented:
I copied over the output from the diff into the summary.

zdevito reviewed (commented):
Code looks good, but the test failure looks legit as well. Ready to go once the tests pass.

salexspb committed:
Profiler: add Self CPU Time Total, CPU time total and other general improvements (#19378)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/19378

Function profile events are typically nested. In this diff I
add parent child relationship to the intervals. This way we can
attribute self time easily. As a result, user printing a table from a
profiler trace gets self cpu time.

This diff doesn't try to address CUDA self time as CUDA kernels are
already getting special care in the profiler.

There are also some other minor improvements. Like reporting total CPU
time spent, reversed sorting, aggregated data after the table,
etc.

There is a new unit test added which tests more functionality than
previous profiler test

Differential Revision: D14988612

fbshipit-source-id: 996eaecf58dc270307d81ba35cdee47d7a9a9198

apaszke commented on torch/autograd/profiler.py:
```diff
@@ -1,30 +1,82 @@
 import torch
 
-
-class range(object):
```
This was a user-facing object, and shouldn't have been deleted. It's the only way to actually insert ranges from Python, which is useful.
